client/state/data-layer/wpcom/read/tags/index.js

```diff
@@ -32,42 +35,30 @@ export function requestTags( store, action, next ) {
 	next( action );
 }
 
-/**
- * Normalize response from the api so whether we get back a single tag or a list of tags
- * we always pass forward a list
- * @param  {Tag|Tags} apiResponse api response from the tags endpoint
- * @return {Tags}             An object containing list of tags
- */
-function fromApi( apiResponse ) {
-	let tags;
-	if ( apiResponse.tag )	 {
-		tags = [ apiResponse.tag ];
-	} else if ( apiResponse.tags ) {
-		tags = map( apiResponse.tags, tag => ( { ...tag, is_following: true } ) );
-	} else {
-		if ( process.env.NODE_ENV === 'development' ) {
-			throw new Error( 'bad api response for /read/tags' );
-		}
-		tags = [];
+export function receiveTagsSuccess( store, action, next, apiResponse ) {
+	let tags = fromApi( apiResponse );
+	// if from the read/tags api, then we should add isFollowing=true to all of the tags
+	if ( apiResponse.tags ) {
+		tags = map( tags, tag => ( { ...tag, isFollowing: true } ) );
 	}
 
-	return { tags };
-}
-
-export function receiveTagsSuccess( store, action, next, apiResponse ) {
 	store.dispatch( receiveTags( {
-		payload: fromApi( apiResponse ),
-		error: false
+		payload: tags,
+		resetFollowingData: !! apiResponse.tags
 	} ) );
 }
 
 export function receiveTagsError( store, action, next, error ) {
-	store.dispatch( receiveTags( {
-		payload: error,
-		error: true
-	} ) );
+	store.dispatch( errorNotice( 'Could not fetch the tag' ) );
+	if ( process.env.NODE_ENV === 'development' ) {
+		throw new Error( error );
+	}
```

Comment:
> how are you planning on preventing us from re-requesting them and spamming the api server?

handle it in the HTTP pipeline with a failure policy description. it seems like passing an empty list of tags on network failure is misleading because it tells Calypso that there is a successful notion that there are no tags. a failure is just that - a failure - and shouldn't indicate any positive notion of the knowledge of what tags really exist

> how are you planning on preventing us from re-requesting them and spamming the api server?

seems like this should be handled in the pipeline.

```js
{
	…,
	failurePolicy: {
		type: EXPONENTIAL_RETRY,
		maxRetries: 3,
		initialDelay: 500,
	}
}
```

Calypso should know when to stop trying